### ISSUE TITLE
spaCy analyzer

### ISSUE THREAD

**Issue by osma** [NatLibFi/Annif#374]
We could add an analyzer based on spaCy. It would enable support for some new languages and possibly also give better results for eg English and German than the current snowball analyzer.

Getting the full benefit of spaCy may require some internal API changes because it is more object oriented than NLTK and processes whole sentences instead of just individual words, taking some of the context into account.

This would be an optional feature as spaCy is implemented as a native code extension, not just pure Python.

**Comment by osma** [comment]
spaCy also has some support for document (text) categorization, both multiclass and multilabel. This could be supported as an Annif backend: https://spacy.io/api/textcategorizer

**Comment by osma** [comment]
Starting to look at this more seriously...

One immediate issue that comes up is that spacy (3.1.2) depends on [typer](https://github.com/tiangolo/typer) (0.3.2) which at the moment depends on click==7.1.2, while Annif depends on click==8.0.1 since PR #499 was merged just before the 0.53 release.

Now this appears to be fixed with [typer 0.4.0](https://github.com/tiangolo/typer/releases/tag/0.4.0) that was released yesterday(!) and supports Click 8, but it won't help until there's a newer version of spaCy available that upgrades the typer dependency.

For now we may have to downgrade back to Click 7.1.2, which probably isn't a problem for Annif since I don't think we've started using any Click 8 features yet.
